src/Data/Variant/Gen.purs

```diff
@@ -0,0 +1,176 @@
+module Data.Variant.Gen where
+
+import Prelude
+
+import Control.Monad.Gen (class MonadGen, frequency, oneOf)
+import Data.List as L
+import Data.NonEmpty (NonEmpty, (:|))
+import Data.Symbol (class IsSymbol)
+import Data.Tuple (Tuple(..))
+import Data.Variant (Variant, inj)
+import Prim.Row as Row
+import Prim.RowList as RL
+import Record as Record
+import Type.Proxy (Proxy(..))
+
+-- Generate a `Variant` with uniform probability given a record whose
+-- generators' labels correspond to the `Variant`'s labels
+--
+-- ```purs
+-- genMyVariant :: Gen (Variant (id :: String, unit :: Unit))
+-- genMyVariant =
+--   genVariantUniform
+--     { id: genAlphaString
+--     , unit: pure unit
+--     }
+-- ```
+--
+-- Inspired by https://github.com/JordanMartinez/purescript-veither/blob/9fb195b7bd97cfcf21bbd389b55002ca458f1bb4/src/Data/Veither.purs#L461
+genVariantUniform
+  :: forall genRows rowList variantRows m
+   . MonadGen m
+  => RL.RowToList genRows rowList
+  => GenVariantUniformNonEmptyList rowList genRows variantRows m
+  => Record genRows
+  -> m (Variant variantRows)
+genVariantUniform = oneOf <<< buildGenVariantUniformNonEmptyList (Proxy :: Proxy rowList)
+
+-- For what functions we can use `NonEmpty List ...` as input? For now - only for `oneOf`
+class
+  GenVariantUniformNonEmptyList (rowList :: RL.RowList Type) (genRows :: Row Type) (variantRows :: Row Type) (m :: Type -> Type)
+  | rowList -> genRows variantRows m
+  where
+  buildGenVariantUniformNonEmptyList :: Proxy rowList -> Record genRows -> NonEmpty L.List (m (Variant variantRows))
+
+-- Instance for a list with at least one element
+instance
+  ( IsSymbol name
+  , Row.Cons name (m a) genRows' genRows
+  , Row.Cons name a variantRows' variantRows
+  , GenVariantUniformList restRowList genRows variantRows m
+  , MonadGen m
+  ) =>
+  GenVariantUniformNonEmptyList (RL.Cons name (m a) restRowList) genRows variantRows m where
+  buildGenVariantUniformNonEmptyList _ rec =
+    let
+      nameProxy = Proxy :: Proxy name
+      restProxy = Proxy :: Proxy restRowList
+      (generator :: m a) = Record.get nameProxy rec
+      (makeVariant :: m (Variant variantRows)) = map (inj nameProxy) generator
```

Comment:
This mixes type annotation styles. If we aren't going to double up the names, can we settle on "postfix" style like with Proxy? That's what's used everywhere else. If the non-proxy signatures are unnecessary for type-checking, maybe just omit them?